packages/gatsby-plugin-sharp/src/index.js

```diff
@@ -95,60 +134,140 @@ function queueImageResizing({ file, args = {}, reporter }) {
   const prefixedSrc =
     options.pathPrefix + `/static/${digestDirPrefix}` + encodedImgSrc
 
-  // Create job and add it to the queue, the queue will be processed inside gatsby-node.js
-  const job = {
-    args: options,
-    inputPath: file.absolutePath,
-    contentDigest: file.internal.contentDigest,
-    outputDir,
-    outputPath: outputFilePath,
+  return {
+    src: prefixedSrc,
+    outputDir: outputDir,
+    relativePath: outputFilePath,
+    width,
+    height,
+    aspectRatio,
+    options,
   }
+}
 
-  const outputFile = path.join(job.outputDir, job.outputPath)
-  let finishedPromise = Promise.resolve()
+async function createJob(job, { reporter }) {
+  const progressBar = createOrGetProgressBar(reporter)
 
-  if (cachedOutputFiles.has(outputFile)) {
-    finishedPromise = cachedOutputFiles.get(outputFile)
+  if (pendingImagesCounter === 0) {
+    progressBar.start()
   }
 
-  // Check if the output file already exists or already is being created.
-  // TODO: Remove this when jobs api is stable, it will have a better check
-  if (!fs.existsSync(outputFile) && !cachedOutputFiles.has(outputFile)) {
-    // schedule job immediately - this will be changed when image processing on demand is implemented
-    finishedPromise = scheduleJob(
-      job,
-      boundActionCreators,
-      pluginOptions,
-      reporter
-    ).then(res => {
-      cachedOutputFiles.delete(outputFile)
+  const transforms = job.args.operations
+  pendingImagesCounter += transforms.length
+  progressBar.total = pendingImagesCounter
 
-      return res
-    })
-
-    cachedOutputFiles.set(outputFile, finishedPromise)
+  if (boundActionCreators.createJobV2) {
+    await boundActionCreators.createJobV2(job)
+  } else {
+    await scheduleJob(job, boundActionCreators)
   }
 
+  progressBar.tick(transforms.length)
+}
+
+function queueImageResizing({ file, args = {}, reporter }) {
+  const {
+    src,
+    width,
+    height,
+    aspectRatio,
+    relativePath,
+    outputDir,
+    options,
+  } = prepareQueue({ file, args })
+
+  // Create job and add it to the queue, the queue will be processed inside gatsby-node.js
+  const finishedPromise = createJob(
+    {
+      name: IMAGE_PROCESSING_JOB_NAME,
+      inputPaths: [file.absolutePath],
+      outputDir,
+      args: {
+        operations: [
+          {
+            outputPath: relativePath,
+            args: options,
+          },
+        ],
+        pluginOptions: getPluginOptions(),
+      },
+    },
+    { reporter }
+  ).catch(err => {
+    reporter.panic(err)
```

Comment:
This `catch` block here keeps everything in the scope alive.

Because internally Gatsby can decide that it already have this job, it will not make use of those objects (as we get a lot of "duplicated" jobs), those objects created here - biggest offender is `options` from `prepareQueue` result, will be kept alive even if they are not used.

One thing that seems to work is to do change this to something like this (remove the chained catch in here and move it to its own function so the scope is limited and allow those unused stuff to be garbage collected):

```js
  const finishedPromise = createJob(
    {
      name: IMAGE_PROCESSING_JOB_NAME,
      inputPaths: [file.absolutePath],
      outputDir,
      args: {
        operations: [
          {
            outputPath: relativePath,
            args: options,
          },
        ],
        pluginOptions: getPluginOptions(),
      },
    },
    { reporter }
  )
  reportStuff(finishedPromise, reporter)
```

and top level function like this (limited scope - should keep only promise and reporter alive):

```js
function reportStuff(jobPromise, reporter) {
  jobPromise.catch(err => {
    reporter.panic(err)
  })
}
```

Reply:
Just wanted to mention that it seems like my reasoning I first wrote about why it's kept in memory doesn't seem to be true after making few simple scripts. It does seems like this is somehow causing stuff not to be released, just reasoning is wrong